### PR TITLE
[SPARK-49162][SQL] Push down date_trunc function

### PR DESCRIPTION
**Pull request by IvanK-db**
### What changes were proposed in this pull request?

This PR:
- Adds support for DATE_TRUNC in V2 optimization pushdown
- Consumes this pushdown for Postgres & H2 Connectors

### Why are the changes needed?

Performance improvements.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New integration test in PostgresIntegrationSuite.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala [hunk 1]
```diff
@@ -47,7 +47,10 @@ private case class PostgresDialect()
   private val supportedAggregateFunctions = Set("MAX", "MIN", "SUM", "COUNT", "AVG",
     "VAR_POP", "VAR_SAMP", "STDDEV_POP", "STDDEV_SAMP", "COVAR_POP", "COVAR_SAMP", "CORR",
     "REGR_INTERCEPT", "REGR_R2", "REGR_SLOPE", "REGR_SXY")
-  private val supportedFunctions = supportedAggregateFunctions
+  private val supportedDatetimeFunctions = Set("DATE_TRUNC")
```
Comment (abhiips07) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:50]: `private val supportedAggregateFunctions: Set[String] = Set(
  "MAX", "MIN", "SUM", "COUNT", "AVG",
  "VAR_POP", "VAR_SAMP", "STDDEV_POP", "STDDEV_SAMP",
  "COVAR_POP", "COVAR_SAMP", "CORR", "REGR_INTERCEPT",
  "REGR_R2", "REGR_SLOPE", "REGR_SXY", "DATE_TRUNC"
)`

This approach simplifies the code and keeps everything in one place. Let me know what you think!

Comment (IvanK-db) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:50]: Imo DATE_TRUNC should not be in the supportedAggregateFunctions as it's not an aggregate function

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala [hunk 2]
```diff
@@ -123,4 +131,14 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCT
       )
     }
   }
+
+  test("SPARK-49162: Push down date_trunc function") {
```
Comment (stefanbuk-db) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:135]: This is good! Could we write a few more tests, maybe test some edge cases. We are adding a new pushdown, let's make sure there are no correctness issues.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala [hunk 3]
```diff
@@ -47,7 +47,10 @@ private case class PostgresDialect()
...
+  private val supportedFunctions =
```
Comment (stefanbuk-db) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:51]: Are we only adding this support for Postgres?

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala (hunk 2 above)
Comment (stefanbuk-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:135]: Also, would be good to have some other variable in trunc function other than 'DAY'

**Comment by HyukjinKwon** [comment]
cc @beliefer

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala (hunk 3 above)
Comment (IvanK-db) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:51]: Yes, only Postgres in this PR

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala [hunk 4]
```diff
@@ -123,4 +132,49 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCT
...
+    def testDateTruncPushdown(format: String, expectedResult: Set[Row]): Unit = {
+      val df = sql(
+        s"""
+            SELECT DATE_TRUNC('$format', time), COUNT(*)
+            | FROM $catalogName.datetime_table
+            | GROUP BY 1
```
Comment (stefanbuk-db) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:142]: Thanks for adding new tests! Can we do tests where DATE_TRUNC is a part of the predicate (where DATE_TRUNC = ....) to see if it works when we push down predicates. I am not sure if we do pushdowns in projection.

Comment (IvanK-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:142]: Added tests for that as well

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala [hunk 5]
```diff
@@ -47,7 +47,10 @@ private case class PostgresDialect()
...
+    supportedAggregateFunctions ++
+      supportedDatetimeFunctions
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:53]: Shall we remove supportedDatetimeFunctions by
`private val supportedFunctions = supportedAggregateFunctions ++ Set("DATE_TRUNC")` ?

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala [hunk 6]
```diff
@@ -123,4 +132,81 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCT
       )
     }
   }
+
+  test("SPARK-49162: Push down aggregate date_trunc function") {
+    def testAggregatePushdown(format: String, expectedResult: Set[Row]): Unit = {
+      val df = sql(
+        s"""
+            SELECT DATE_TRUNC('$format', time), COUNT(*)
```
Comment (beliefer) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:140]: Is `DATE_TRUNC` supported by H2 database ?

Comment (IvanK-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:140]: https://www.h2database.com/html/functions.html#date_trunc

Looks like yes, however the original task was adding support only for Postgres for now.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala (hunk 5 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:53]: The aggregate functions is very different from others(string functions, datetime functions and so on) and already defined a lot of aggregate functions.
We can define `supportedDatetimeFunctions` in future if need.

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala (hunk 6 above)
Comment (beliefer) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:140]: I think we should support for H2 dialect first, then other database dialect.

Comment (IvanK-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:140]: Added pushdown for H2 as well

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala (hunk 5 above)
Comment (IvanK-db) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:53]: Done

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala (hunk 4 above)
Comment (beliefer) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:142]: Why introduce aggregate here? we just need test predicates.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala [hunk 7]
```diff
@@ -1599,6 +1599,14 @@ class JDBCV2Suite extends QueryTest with SharedSparkSession with ExplainSuiteHel
       "PushedTopN: ORDER BY [EXTRACT(DAY_OF_YEAR FROM DATE1) ASC NULLS FIRST] LIMIT 1,"
     checkPushedInfo(df9, expectedPlanFragment9)
     checkAnswer(df9, Seq(Row("alex")))
+
+    val df10 = sql("SELECT name FROM h2.test.datetime WHERE " +
+      "DATE_TRUNC('DAY', date1) = date'2022-05-19'")
```
[the remark this answers is not in the document]
Comment (IvanK-db) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala]: Done

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala (hunk 4 above)
Comment (IvanK-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:142]: The original query that was used to point out the function wasn't pushed down was similar to this one so i modeled the test case after it. Should we remove the aggregate test cases?

Comment (beliefer) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:142]: Yes. Let's remove them.

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala [hunk 8]
```diff
@@ -123,4 +132,81 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCT
...
+        s"""
+            SELECT *
+           | FROM $catalogName.datetime_table
+           | WHERE DATE_TRUNC('$format', time) = '$date'
+         """.stripMargin
+      )
+      val filters = df.queryExecution.optimizedPlan.collect {
+        case f: Filter => f
+      }
+      assert(filters.isEmpty)
```
Comment (beliefer) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:193]: Please reuse `checkFilterPushed` here.

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala (hunk 4 above)
Comment (IvanK-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala]: Done

**Comment by IvanK-db** [comment]
@beliefer Kind reminder, all comments have been addressed.

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala [hunk 9]
```diff
@@ -353,7 +353,7 @@ private[v2] trait V2JDBCTest extends SharedSparkSession with DockerIntegrationFu
     }
   }
 
-  private def checkFilterPushed(df: DataFrame, pushed: Boolean = true): Unit = {
+  protected def checkFilterPushed(df: DataFrame, pushed: Boolean = true): Unit = {
```
Comment (beliefer) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:356]: Could you rebase the repository? because the modifier already is protected.

**Review by beliefer — APPROVED** [review]
LGTM. cc @cloud-fan Do you have any other comments?

**Review by stefanbuk-db — APPROVED** [review]

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala [hunk 10]
```diff
@@ -123,4 +131,33 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JDBCT
       )
     }
   }
+
+  test("SPARK-49162: Push down filter date_trunc function") {
```
Comment (milastdbx) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:135]: reading the documentation for pgsql:
https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-TRUNC
and databricks:
https://docs.databricks.com/en/sql/language-manual/functions/date_trunc.html

I see that they support `microseconds` while we support `microsecond`.

Can we test for all our supported precisions ? ALso can we try different casings ?

**Review by milastdbx — CHANGES_REQUESTED** [review]
Lets add more tests for date_trunc precisions.

Also, please update description a bit.
In this PR you are:
 - Adding support for `DATE_TRUNC` in V2 optimization pushdown
 - Consuming this pushdown for Postgres & H2 Connectors

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala [hunk 11]
```diff
@@ -267,6 +267,7 @@ class V2ExpressionBuilder(e: Expression, isPredicate: Boolean = false) extends L
     case _: DateAdd => generateExpressionWithName("DATE_ADD", expr, isPredicate)
     case _: DateDiff => generateExpressionWithName("DATE_DIFF", expr, isPredicate)
     case _: TruncDate => generateExpressionWithName("TRUNC", expr, isPredicate)
+    case _: TruncTimestamp => generateExpressionWithName("DATE_TRUNC", expr, isPredicate)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala:270]: is `DATE_TRUNC` a standard SQL function or widely supported in the industry?

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala (hunk 10 above)
Comment (beliefer) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:135]: Yes. Could we add other precisions for H2 and Postgres? @IvanK-db

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala (hunk 11 above)
Comment (IvanK-db) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala:270]: Not a standard function but it's supported by several databases - Postgres, Redshift, Snowflake, BQ

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala (hunk 10 above)
Comment (IvanK-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/PostgresIntegrationSuite.scala:135]: Added

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala [hunk 12]
```diff
@@ -1601,6 +1603,80 @@ class JDBCV2Suite extends QueryTest with SharedSparkSession with ExplainSuiteHel
...
+    checkPushedInfo(df8, expectedPlanFragment8)
+    checkAnswer(df8, Seq(Row("adam")))
+
+    val df9 = sql("SELECT name FROM h2.test.datetime WHERE " +
+      "DATE_TRUNC('MicroseconD', time1) = timestamp'2024-09-05 11:23:45.000000'")
+    checkFiltersRemoved(df9)
+    val expectedPlanFragment9 =
+      "PushedFilters: [(DATE_TRUNC('MicroseconD', TIME1)) = 1725560625000000]"
+    checkPushedInfo(df9, expectedPlanFragment9)
+    checkAnswer(df9, Seq(Row("adam")))
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala:1677]: Please add test cases for all supported precisions.
Please add the negative test cases, such as: `MicroSecondS`.

Comment (IvanK-db) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala:1677]: Added all precisions. MicroSecondS seems to be a positive test case, looks to be an alias of microsecond so i added another negative test case.

**Review by beliefer — APPROVED** [review]
LGTM.

**Comment by cloud-fan** [comment]
My biggest concern is we translate the Spark `TruncTimestamp` expression to v2 `DATE_TRUNC` function. `DATE_TRUNC` is not in the SQL standard and personally I think it's more reasonable to let the v2 `TRUNC` function accept timestamp inputs. Redshift's TRUNC function also supports timestamp: https://docs.aws.amazon.com/redshift/latest/dg/r_TRUNC.html
